### PR TITLE
Allow exporting schema into multiple files

### PR DESCRIPTION
**Pull request by u12206050**
## Description

When doing `schema snapshot` add `--merged no` as argument and specify the schema path as such `schema/schema.yaml` in order to export the schema as multiple files, one for each collection.

So `directus schema snapshot --merged no schema/schema.yaml -y`

Default for `merged` is yes, as to be backward compatible.

`schema apply` works the same, however it will now check if the snapshot is `partial` meaning in multiple files before applying it to the database.

Ref #16101

## Type of Change

- [ ] Bugfix
- [x] Improvement
- [x] New Feature
- [ ] Refactor / codestyle updates
- [ ] Other, please describe:

## Requirements Checklist

- [ ] New / updated tests are included
- [x] All tests are passing locally
- [ ] Performed a self-review of the submitted code

If adding a new feature:

- [x] Documentation was added/updated. PR: https://github.com/directus/docs/pull/267

### CODE REVIEW

**Comment by u12206050** [comment]
Another benefit with this approach is being able to more easily select what changes you want to commit to GIT instead of trying to cherry pick from a large schema file.

**Comment by u12206050** [comment]
SchemaUnmerged

**Comment by tenebrius** [comment]
Another solution to this would be save only the new changes. similar to database migrations
e.g

Create schema migrations
 - if no migrations, create a migration with full snapshot
- else
  - get all existing migration. 
  - compile into one big snapshot. 
  - get  changes with current database snapshot. and save only the changes into a new migration file

Apply schema migrations
- get all existing migrations
- compile into one big snapshot. 
- compare with current snapshot and apply the changes

**Comment by azrikahar** [comment]
Apart from the original discussion #16100 and the other past discussions linked here: https://github.com/directus/directus/discussions/16100#discussioncomment-3947104, I think we should take into account the upcoming schema endpoints #14713 in terms of feature parity between CLI and API.

**Comment by u12206050** [comment]
In a team setup with 20+ collections, the schema file is large and often has conflicts this PR could really solve it and make it more simpler to maintain.
It is fully backwards compatible, is there anyway we could get this in Production faster (without starting our own fork:)
@azrikahar 

Thanks in advance.

**Comment by paescuj** [comment]
@u12206050 I haven't had a chance to look at your PR in detail yet, but I can say that it looks very interesting.
A short feedback for now:
- For me, the name of the option doesn't seem that meaningful. I'd probably take the opposite path and use something like `--split` (that is, when the option is given it would act like `yes|true` instead of `--merged no`).
- > `schema apply` works the same, however it will now check if the snapshot is `partial` meaning in multiple files before applying it to the database.
  
  What's the motivation behind not depending on the option here as well? Wouldn't this simplify the logic considerably?

File: api/src/cli/commands/schema/apply.ts [hunk 1]
```diff
@@ -26,13 +26,43 @@ export async function apply(snapshotPath: string, options?: { yes: boolean; dryR
 
 	let snapshot: Snapshot;
 
-	try {
+	async function loadSnapshot<T extends Snapshot | SnapshotCollection>(filename: any): Promise<T> {
```
Comment (azrikahar) [inline api/src/cli/commands/schema/apply.ts:29]:
```suggestion
	async function loadSnapshot<T extends Snapshot | SnapshotCollection>(filename: string): Promise<T> {
```

File: api/src/cli/commands/schema/apply.ts [hunk 2]
```diff
@@ -26,13 +26,43 @@ export async function apply(snapshotPath: string, options?: { yes: boolean; dryR
...
+			snapshot.collections = [];
+			snapshot.fields = [];
+			snapshot.relations = [];
+
+			const { dir, name, ext } = path.parse(path.resolve(process.cwd(), filename));
+			const files = await fs.readdir(dir);
+			const regex = new RegExp(`${name}-(.+)${ext}$`);
+
+			for (const file of files) {
+				if (file.match(regex) !== null) {
```
Comment (azrikahar) [inline api/src/cli/commands/schema/apply.ts:52]:
```suggestion
				if (file.match(regex)) {
```

File: api/src/cli/commands/schema/snapshot.ts [hunk 3]
```diff
@@ -8,13 +8,11 @@ import { dump as toYaml } from 'js-yaml';
...
 	snapshotPath?: string,
-	options?: { yes: boolean; format: 'json' | 'yaml' }
+	options?: { yes: boolean; format: 'json' | 'yaml'; split: 'yes' | 'no' }
 ): Promise<void> {
 	const database = getDatabase();
 
-	try {
-		const snapshot = await getSnapshot({ database });
-
+	async function saveSnapshot(snapshot: any, suffix = '') {
```
Comment (azrikahar) [inline api/src/cli/commands/schema/snapshot.ts:15]: thoughts on making suffix an optional param (`suffix?: string`) rather than default to `''`?

**Review by azrikahar — CHANGES_REQUESTED** [review]
> It is fully backwards compatible, is there anyway we could get this in Production faster (without starting our own fork:)

One of the concern for me would be this change may be superseded by a future change (hence the keys within each split files may change), but I fully understand how this solution does alleviate said pain point for the time being. Currently there are other tasks that are slated for the coming weeks and will need to be prioritized over this, but I'll see what I can do 👍

File: api/src/cli/index.ts [hunk 4]
```diff
@@ -92,7 +92,7 @@ export async function createCli(): Promise<Command> {
 		.description('Create a new Schema Snapshot')
 		.option('-y, --yes', `Assume "yes" as answer to all prompts and run non-interactively`, false)
 		.addOption(new Option('--format <format>', 'JSON or YAML format').choices(['json', 'yaml']).default('yaml'))
-		.addOption(new Option('--split <split>', 'Export as a single schema file').choices(['yes', 'no']).default('no'))
+		.addOption('-s, --split', 'Export as a single schema file', false)
```
Comment (azrikahar) [inline api/src/cli/index.ts:95]: Would it possible to update this description? It feels like it's describing the opposite from my current interpretation, but I could be misunderstanding it.

**Review by azrikahar — COMMENTED** [review]
@u12206050 Thanks for the swift updates! Would you mind opening a PR over in `directus/docs` to include this in https://docs.directus.io/self-hosted/cli.html as well? Just so we can better guide users how to effectively use this useful flag 👍 Really appreciate the help!

File: api/src/cli/index.ts (hunk 4 above)
Comment (u12206050) [reply api/src/cli/index.ts]: Ah right, you are right

**Comment by u12206050** [comment]
Docs updated: https://github.com/directus/docs/pull/267

**Comment by u12206050** [comment]
My schema file is now 24000 lines, could really use this addition for easier management. @azrikahar  anything else out standing or in the way for this to get merged?

**Comment by azrikahar** [comment]
> My schema file is now 24000 lines, could really use this addition for easier management. @azrikahar anything else out standing or in the way for this to get merged?

There is currently work that has a higher priority and unfortunately we lack the time to review this pull request with the necessary attention. As this is an unplanned feature, it requires extra caution as we need to be sure if we can and want to support it in this form over a longer period of time. Thank you for the understanding!

**Comment by rijkvanzanten** [comment]
Heya, thanks for your PR! Our small team has been attempting to work through a backlog of external contributions for the better part of a year, but in that time many have become too stale to merge or the change is no longer desired/relevant.. 
We've updated the [contributing guidelines](https://github.com/directus/directus/blob/main/contributing.md#pull-requests) and are working on [making more resources available](https://github.com/directus/directus/discussions/17977) to make sure that doesn’t happen again 😬 In the meantime, we'll have to close this PR for now..

**Comment by ignasigarcia** [comment]
Hey, our team would greatly benefit from this feature.

Currently we are experiencing merging issues with the schema and that would alleviate if not fix our issues.

Happy to see that this is being considered.

Thanks!

**Pull request closed without merging**